spec/rester/utils/circuit_breaker_spec.rb

```diff
@@ -0,0 +1,490 @@
+          end
+        end
+
+        context 'without reaching threshold' do
+          after { (threshold - 1).times { failure_call } }
+
+          it 'should not call callback' do
+            expect(callback).not_to receive :call
+          end
+        end
+
+        context 'with threshold reached' do
+          after { threshold.times { failure_call } }
+
+          it 'should not call callback' do
+            expect(callback).not_to receive :call
+          end
+        end
+
+        context 'with threshold surpassed' do
+          after { (threshold * 3).times { failure_call } }
+
+          it 'should not call callback' do
+            expect(callback).not_to receive :call
+          end
+        end
+
+        context 'with successful call after threshold reached' do
+          let(:retry_period) { 0.001 }
+          after do
+            threshold.times { failure_call }
+            let_retry_period_pass
+            success_call
+          end
+
+          it 'should call callback once' do
+            expect(callback).to receive(:call).once
+          end
+        end # with successful call after threshold reached
+
+        context 'with many successful call after threshold reached' do
+          let(:retry_period) { 0.001 }
+          after do
+            threshold.times { failure_call }
+            let_retry_period_pass
+            (threshold * 3).times { success_call }
+          end
+
+          it 'should call callback once' do
+            expect(callback).to receive(:call).once
+          end
+        end # with many successful call after threshold reached
+
+        context 'with circuit opened and closed 5 times' do
+          let(:retry_period) { 0.001 }
+          after do
+            5.times {
+              threshold.times { failure_call }
+              let_retry_period_pass
+              success_call
+            }
+          end
+
+          it 'should call callback 5 times' do
+            expect(callback).to receive(:call).exactly(5).times
+          end
+        end # with circuit opened and closed 5 times
+      end # #on_close
+
+      describe '#call' do
+        subject { breaker.call }
+
+        context 'with successful call' do
+          before { allow(breaker.block).to receive(:call).and_return 'hi' }
+
+          it 'should return retval of block' do
+            is_expected.to eq 'hi'
+          end
+        end # with successful call
+
+        context 'with failed call' do
+          before { allow(breaker.block).to receive(:call).and_raise 'hi' }
+
+          it 'should raise exception raised by block' do
+            expect { subject }.to raise_error RuntimeError, 'hi'
+          end
+
+          context 'with threshold reached' do
+            before { threshold.times { failure_call } }
+
+            it 'should raise CircuitOpenError' do
+              expect { subject }.to raise_error CircuitBreaker::CircuitOpenError
+            end
+          end # with threshold reached
```

Comment:
Add test to check that we're not calling the block once the threshold has been reached, that we call the block once on retry and that we start calling the block again after recovering.